includes/rest-api/Controllers/Version3/class-wc-rest-orders-controller.php

```diff
@@ -34,30 +34,45 @@ class WC_REST_Orders_Controller extends WC_REST_Orders_V2_Controller {
 	 * @return bool
 	 */
 	protected function calculate_coupons( $request, $order ) {
-		if ( ! isset( $request['coupon_lines'] ) || ! is_array( $request['coupon_lines'] ) ) {
+		if ( ! isset( $request['coupon_lines'] ) ) {
 			return false;
 		}
 
-		// Remove all coupons first to ensure calculation is correct.
-		foreach ( $order->get_items( 'coupon' ) as $coupon ) {
-			$order->remove_coupon( $coupon->get_code() );
-		}
+		// Validate input and at the same time generate the coupon objects to apply.
+
+		$coupons   = array();
+		$discounts = new WC_Discounts( $order );
 
 		foreach ( $request['coupon_lines'] as $item ) {
-			if ( is_array( $item ) ) {
-				if ( ! empty( $item['id'] ) ) {
-					throw new WC_REST_Exception( 'woocommerce_rest_coupon_item_id_readonly', __( 'Coupon item ID is readonly.', 'woocommerce' ), 400 );
-				}
+			if ( ! empty( $item['id'] ) ) {
+				throw new WC_REST_Exception( 'woocommerce_rest_coupon_item_id_readonly', __( 'Coupon item ID is readonly.', 'woocommerce' ), 400 );
+			}
 
-				if ( empty( $item['code'] ) ) {
-					throw new WC_REST_Exception( 'woocommerce_rest_invalid_coupon', __( 'Coupon code is required.', 'woocommerce' ), 400 );
-				}
+			if ( empty( $item['code'] ) ) {
+				throw new WC_REST_Exception( 'woocommerce_rest_invalid_coupon', __( 'Coupon code is required.', 'woocommerce' ), 400 );
+			}
 
-				$results = $order->apply_coupon( wc_clean( $item['code'] ) );
+			$coupon_code  = wc_format_coupon_code( wc_clean( $item['code'] ) );
+			$coupon       = new WC_Coupon( $coupon_code );
+			$check_result = $discounts->is_coupon_valid( $coupon );
```

Comment:
so this check here has an unintended side effect, which is that `is_coupon_valid` also checks for the coupon limits usage.

this means that if I have a coupon already applied to the order, whose limit is already been reached, and try to apply it again using the REST API, it will throw an error. this is a bit different from earlier behavior where we did allow sending the same coupon which is already applied with limit exhausted. it worked earlier because we would remove the coupon, which adds back the limit, and then apply it back again.

i think we should exclude the already applied coupon from this check to make it compatible with existing behavior. wdyt?